### PR TITLE
[Faster Transformer][Ready for review] Support unified transformer

### PR DESCRIPTION
**Pull request by FrostML**
### PR types

New features

### PR changes

Others

### Description

Support unified transformer.
【Can be Reviewed】

### CODE REVIEW

File: examples/dialogue/unified_transformer/utils.py [hunk 1]
```diff
@@ -230,3 +233,25 @@ def select_response(ids,
...
+    for pred in ids:
+        pred_token_ids, pred_tokens = post_process_response(pred, tokenizer)
+        num_token = len(pred_token_ids)
+        if keep_space:
+            response = " ".join(pred_tokens)
+        else:
+            response = "".join(pred_tokens)
+
+        in_turn_repetition = get_in_turn_repetition(
+            pred_tokens, True) or get_in_turn_repetition(pred_token_ids)
```
Comment (guoshengCS) [inline examples/dialogue/unified_transformer/utils.py]: 这里`in_turn_repetition `没有使用就去掉吧，另外是否考虑在utils.py里将`select_response_without_scores`和`select_response`合并

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cc [hunk 2]
```diff
@@ -78,7 +78,8 @@ std::vector<paddle::Tensor> UnifiedDecodingForward(
     output_dims = {max_len, batch_size, beam_size};
     parent_ids_dims = output_dims;
   } else if (decoding_strategy == "topk_sampling" ||
-             decoding_strategy == "topp_sampling") {
+             decoding_strategy == "topp_sampling" ||
+             decoding_strategy == "sampling") {
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cc:82]: 感觉就topk_sampling/topp_sampling或者就像python端一样用sampling配合topk、topp区分也可以，同时支持反倒有些混乱且意义不大

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h [hunk 3]
```diff
@@ -505,6 +505,8 @@ class TransformerBeamsearch {
         // TODO(): repeat penalty vertification.
         apply_penalties_Launcher<float>(step,
                                         logits_buf_,
+                                        decoding_params.embedding_bias,
+                                        finished_buf_,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: penalty通常还是作用在log_prob上的，函数中相应的这个参数名也是`log_prob`

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h [hunk 4]
```diff
@@ -611,7 +608,7 @@ class TransformerBeamsearch {
           args_.batch_size_,
           args_.beam_width_,
           args_.hidden_units_,
-          step,
+          step + args_.start_len_,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: 这里是否要和上面当前步使用的`int kv_cache_id = step & 0x1;`一致，也要使用step

File: paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h [hunk 5]
```diff
@@ -520,6 +518,8 @@ class TransformerSampling {
         // TODO(): repeat penalty vertification.
         apply_penalties_Launcher(step,
                                  logits_buf_,
+                                 decoding_params.embedding_bias_T,
+                                 finished_buf_,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h:526]: 同上，注意penalty作用在log_prob上。不行就先去掉penalty，去掉的话注意下面去掉的两个logits加bias再加回去。

File: paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cc (hunk 2 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/src/fusion_unified_decoding_op.cc:82]: `topk_sampling`/`topp_sampling` 是为了和之前封装的 transformer，gpt 一致，`sampling` 是为了和 python 一致，考虑是 python 使用的时候，在传入 `args.decoding_strategy = "sampling"` 的时候不会额外区分 `topk`/`topp`，额外再转一次显得 API 封装比较冗余，后续统一都改成 sampling 吧，也和 python 统一。

File: examples/dialogue/unified_transformer/utils.py (hunk 1 above)
Comment (FrostML) [reply examples/dialogue/unified_transformer/utils.py]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h (hunk 3 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: Fixed. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h (hunk 4 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: Fixed. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h (hunk 5 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h:526]: Fixed. Thanks.

File: paddlenlp/ops/faster_transformer/transformer/faster_transformer.py [hunk 6]
```diff
@@ -478,3 +480,138 @@ def save_resources(self, tokenizer, path):
...
+                 decoding_type_id=1):
+        super(FasterUnifiedTransformer, self).__init__()
+        self._model = model
+        self._decoding_strategy = decoding_strategy
+        self.bos_token_id = model.bos_token_id
+        self.pad_token_id = model.pad_token_id
+        self.eos_token_id = model.eos_token_id
+        self.unk_token_id = model.unk_token_id
+        self.vocab_size = model.lm_head.decoder_bias.shape[0]
+        self.logits_mask = self.generate_logits_mask(use_fp16_decoding)
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/transformer/faster_transformer.py:500]: 当前这样存在非参数的导出模型会有问题吗，这个FasterUnifiedTransformer当前还没有支持静态图或者动转静导出是吗

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h [hunk 7]
```diff
@@ -470,134 +474,106 @@ class TransformerBeamsearch {
       check_cuda_error(cudaGetLastError());
 #endif
 
+      float alpha_ = (DataType_)1.0f;
+      float beta_ = (DataType_)0.0f;
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: 注意编码规范，_一般是成员变量使用

File: paddlenlp/ops/faster_transformer/transformer/faster_transformer.py [hunk 8]
```diff
@@ -479,3 +481,138 @@ def save_resources(self, tokenizer, path):
...
             tokenizer.save_resources(path)
+
+
+class FasterUnifiedTransformer(UnifiedTransformerPretrainedModel):
+    def __init__(self,
+                 model,
+                 decoding_strategy="sampling",
+                 decoding_lib=None,
+                 use_fp16_decoding=False,
+                 decoding_type_id=1):
```
Comment (guoshengCS) [inline paddlenlp/ops/faster_transformer/transformer/faster_transformer.py]: `decoding_type_id`是和输入中token_type_ids一样的作用吗，相比当前非加速版本的话当前这样是否无法支持不同batch有不同的token_type_ids呢

File: paddlenlp/ops/patches/FasterTransformer/cuda/transformer_decoding_kernels.cu [hunk 9]
```diff
@@ -279,6 +292,38 @@ void apply_penalties_Launcher(int step,
...
+void update_KV_cache_kernelLauncher(T** key_cache,
+                                    T** value_cache,
+                                    const int* beam_ids,
+                                    const int batch_size,
+                                    const int beam_width,
+                                    const int hidden_dim,
+                                    const int step,
+                                    const int start_len,
+                                    const int cache_size,
+                                    const int decoder_layers,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/transformer_decoding_kernels.cu:305]: 如果只是比FT原有的update_KV_cache_kernelLauncher多一个start_len的话直接还调用原来的呢，只是step参数传入`step+start_len`

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h [hunk 10]
```diff
@@ -381,8 +381,11 @@ class TransformerBeamsearch {
 #endif
       }
 
-      float alpha = (float)1.0f;
-      float beta = (float)0.0f;
+      int cublasAlgo = (sizeof(DataType_) == 2) ? CUBLAS_GEMM_DEFAULT_TENSOR_OP
+                                                : CUBLAS_GEMM_DEFAULT;
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: 这里还是和上面的cublasAlgo_设置没区别吧
```
      if (Traits_::OpType == OperationType::FP32) {
        cublasAlgo_[0] = CUBLAS_GEMM_DEFAULT;
      } else {
        cublasAlgo_[0] = CUBLAS_GEMM_DEFAULT_TENSOR_OP;
      }
```

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h [hunk 11]
```diff
@@ -470,134 +474,106 @@ class TransformerBeamsearch {
...
-                      decoding_params.embedding_bias,
-                      args_.end_id_,
-                      finished_buf_,
-                      m,
-                      n,
-                      decoding_params.stream);
+
+      // TODO(): repeat penalty vertification.
+      apply_penalties_Launcher<float>(step,
+                                      logits_buf_,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h:530]: beam search的length penalty作用于cum_log_prob上，这个还要注意下

File: paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h [hunk 12]
```diff
@@ -564,6 +560,23 @@ class TransformerSampling {
                                n,
                                decoding_params.stream);
 
+        // TODO(): repeat penalty vertification.
+        apply_penalties_Launcher(step,
+                                 logits_buf_,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h:562]: 这里的penalty去掉吧，按照generate生成API中的话sampling这个是没有length penalty的

**Review by smallv0221 — DISMISSED** [review]
LGTM for generate API

File: paddlenlp/ops/faster_transformer/transformer/faster_transformer.py (hunk 6 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/transformer/faster_transformer.py:500]: 已支持静态图。

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h (hunk 7 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: Done. Thanks.

File: paddlenlp/ops/faster_transformer/transformer/faster_transformer.py (hunk 8 above)
Comment (FrostML) [reply paddlenlp/ops/faster_transformer/transformer/faster_transformer.py]: 已支持从 forward 传入。

File: paddlenlp/ops/patches/FasterTransformer/cuda/transformer_decoding_kernels.cu (hunk 9 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/cuda/transformer_decoding_kernels.cu:305]: 如上次 review comments，保留现有修改。

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h (hunk 10 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h (hunk 11 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h:530]: Fixed. Thanks. 
这段重构了 logits 的计算，把 `logits_mask` 和 `penalty` 分开了。

File: paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h (hunk 12 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_sampling.h:562]: Done. Thanks.

File: examples/dialogue/unified_transformer/infer.py [hunk 13]
```diff
@@ -98,15 +111,24 @@ def infer(args):
...
             print('step %d - %.3fs/step' %
                   (step, total_time / args.logging_steps))
             total_time = 0.0
-        results = select_response(ids, scores, tokenizer, args.max_dec_len,
-                                  args.num_return_sequences)
+
+        if args.faster:
+            ids = output
+            print(ids)
+            exit()
```
Comment (guoshengCS) [inline examples/dialogue/unified_transformer/infer.py]: 注意去掉

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h [hunk 14]
```diff
@@ -325,6 +325,7 @@ class TransformerBeamsearch {
         check_cuda_error(cudaGetLastError());
 #endif
 
+        decoder_->initialize_stream(decoding_params.stream);
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h:326]: 这里是什么作用呢，不是已经在decoder的initialize中设置过了吗

**Review by guoshengCS — DISMISSED** [review]
LGTM

File: examples/dialogue/unified_transformer/infer.py (hunk 13 above)
Comment (FrostML) [reply examples/dialogue/unified_transformer/infer.py]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h (hunk 14 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_beamsearch.h:326]: 后面使用的是 `decoder_` 的成员函数，所以需要在 `post-norm` 情况下，提前初始化 `stream`。仅初始化 `stream` 即可，所以新增一个简单的函数。

**Review by guoshengCS — APPROVED** [review]
